### ISSUE TITLE
Turning a Report that has been downcast to std::error::Error, back into a Report

### ISSUE THREAD

**Issue by Mythra** [yaahc/color-eyre#79]
Hey Jane :wave: ,

tl;dr library takes my function that returns a `Report`, and returns gives me a `dyn std::error::Error`, how can I get it back to a `Report` to add extra notes onto it?

I've been doing some work with Hyper, and I've run into a situation I'm not quite sure how to handle. I'm working with hyper, and an HTTP Client that supports everything I need it to support. In a custom client I have some code that looks like this:

```rust
fn read_request(&self, req: hyper::Request) -> Result<Body> {
  // setup code
  let resp_result = self.underlying_client.request(req).await;
  if let Err(resp_err) = resp_result {
    return Err(hyper_to_eyre(/* snipped other pieces of context */, resp_err));
  }
  // do things with response.
}
```

The `.request()` call ends up calling my custom connector which the relevant bit of context looks like:

```rust
let resp = other_conn.call(req).await;
if let Err(resp_err) = resp {
  return Err(tcp_stream_to_eyre(resp_err));
}
Ok(EtherStream::Tcp(resp.unwrap()))
```

If you notice in the error case it is capable of returning a custom `Report` error which already has _some_ context in it, but not all of the context I want. Unfortunately the error I get in `read_request` is marked as a `dyn std::error::Error`, because hyper can throw several of it's own errors as well in this same codepath (my connector call isn't the only thing happening here).

How can I turn this back into a `Report`, or something that keeps all the context of the original report, but still allows me to add extra notes?

**Comment by Mythra** [comment]
FWIW this is currently not possible, and while there is work ongoing in the meantime if someone runs into a similar error I got around this by not returning a `Report`  (in the initial error that gets `Into<Box<std::errror::Error>>`), but instead returning a type I wrote which carries the report through without losing the context.

The type is the below:

```rust
pub struct EyreSpyASmuggler {
	inner: Report,
}

impl EyreSpyASmuggler {
	/// Construct a new temporary error holder.
	pub fn new(reportable: Report) -> Self {
		Self { inner: reportable }
	}

	/// Consume an error holder to get the report back out of it.
	pub fn consume(self) -> Report {
		self.inner
	}
}

impl std::fmt::Display for EyreSpyASmuggler {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		write!(f, "{}", self.inner)
	}
}

impl std::fmt::Debug for EyreSpyASmuggler {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		write!(f, "{:?}", self.inner)
	}
}

impl std::error::Error for EyreSpyASmuggler {
	fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
		None
	}
}
```

This prints out exactly like a report would incase someone forgets to unwrap it, or use it directly. However in the case mentioned above I can just consume it, and use it like so. The code originally producing the error looks like this:

```rust
let resp = other_conn.call(req).await;
if let Err(resp_err) = resp {
  return Err(EyreSpyASmuggler::new(tcp_stream_to_eyre(resp_err)));
}
Ok(EtherStream::Tcp(resp.unwrap()))
```

Then when handling it, to get the report back I simply downcast, and consume:

```rust
if cause.is::<EyreSpyASmuggler>() {
  let as_report: Report = cause.downcast::<EyreSpyASmuggler>().unwrap().consume();
  // ... Do stuff with report.
} else {
  eyre!("Unknown Cause!")
}
```
